tests/nonbonded/test_nonbonded_pair_list.py

```diff
@@ -0,0 +1,166 @@
+        )
+        return jax.numpy.sum(scales[:, 1] * vdW + scales[:, 0] * electrostatics)
+
+    return wrapped
+
+
+@pytest.mark.parametrize("lamb", [0.0, 0.1])
+@pytest.mark.parametrize("beta", [2.0])
+@pytest.mark.parametrize("cutoff", [1.1])
+@pytest.mark.parametrize("precision,rtol,atol", [(np.float64, 1e-8, 1e-8), (np.float32, 1e-4, 5e-4)])
+@pytest.mark.parametrize("ixn_group_size", [2, 33, 231])
+def test_nonbonded_pair_list_correctness(
+    ixn_group_size,
+    precision,
+    rtol,
+    atol,
+    cutoff,
+    beta,
+    lamb,
+    example_nonbonded_params,
+    example_conf,
+    example_box,
+    rng: np.random.Generator,
+):
+    "Compares with jax reference implementation."
+
+    num_atoms, _ = example_conf.shape
+
+    # randomly select 2 interaction groups and construct all pairwise interactions
+    atom_idxs = rng.choice(
+        num_atoms,
+        size=(
+            2,
+            ixn_group_size,
+        ),
+        replace=False,
+    ).astype(np.int32)
+
+    pair_idxs = np.stack(np.meshgrid(atom_idxs[0, :], atom_idxs[1, :])).reshape(2, -1).T
+    num_pairs, _ = pair_idxs.shape
+
+    scales = rng.uniform(0, 1, size=(num_pairs, 2))
+
+    lambda_plane_idxs = rng.integers(-2, 3, size=(num_atoms,), dtype=np.int32)
+    lambda_offset_idxs = rng.integers(-2, 3, size=(num_atoms,), dtype=np.int32)
+
+    ref_potential = make_ref_potential(pair_idxs, scales, lambda_plane_idxs, lambda_offset_idxs, beta, cutoff)
+    test_potential = NonbondedPairList(pair_idxs, scales, lambda_plane_idxs, lambda_offset_idxs, beta, cutoff)
+
+    GradientTest().compare_forces(
+        example_conf,
+        example_nonbonded_params,
+        example_box,
+        lamb,
+        ref_potential,
+        test_potential,
+        precision=precision,
+        rtol=rtol,
+        atol=atol,
+    )
+
+
+@pytest.mark.parametrize("lamb", [0.0, 0.1])
+@pytest.mark.parametrize("beta", [2.0])
+@pytest.mark.parametrize("cutoff", [1.1])
+@pytest.mark.parametrize("precision,rtol,atol", [(np.float64, 1e-8, 1e-8), (np.float32, 1e-4, 5e-4)])
+@pytest.mark.parametrize("ixn_group_size", [2, 33, 231])
+def test_nonbonded_pair_list_interpolated_correctness(
+    ixn_group_size,
+    precision,
+    rtol,
+    atol,
+    cutoff,
+    beta,
+    lamb,
+    example_nonbonded_params,
+    example_conf,
+    example_box,
+    rng: np.random.Generator,
+):
+    "Compares with jax reference implementation, with parameter interpolation."
+
+    num_atoms, _ = example_conf.shape
+
+    params_initial = example_nonbonded_params
+    params_final = params_initial + rng.normal(0, 0.01, size=params_initial.shape)
+    params = np.concatenate((params_initial, params_final))
+
+    # randomly select 2 interaction groups and construct all pairwise interactions
+    atom_idxs = rng.choice(
+        num_atoms,
+        size=(
+            2,
+            ixn_group_size,
+        ),
+        replace=False,
+    ).astype(np.int32)
```

Comment:
Question: does this generate inputs where `len(set(atom_idxs[0]).intersection(atom_idxs[1])) > 0`?

If so, it's good that this case is handled... Actually, does the code skip or include spurious self-interactions `(i, i)`? If these are included, they could possibly cause overflows or other issues.

In any case, it may be better to test this case separately from the standard case where `atom_idxs[0]` and `atom_idxs[1]` are disjoint.

Reply:
> Question: does this generate inputs where len(set(atom_idxs[0]).intersection(atom_idxs[1])) > 0?

No, the intersection should be empty (`replace=False` ensures this).

> Actually, does the code skip or include spurious self-interactions (i, i)?

The `NonbondedPairList` constructor will [throw a runtime error](https://github.com/proteneer/timemachine/blob/a4a06d93df52536c115017f9b6bc2c3c329c3935/timemachine/cpp/src/nonbonded_pair_list.cu#L30-L37) if there are self interactions (tested [here](https://github.com/proteneer/timemachine/blob/a4a06d93df52536c115017f9b6bc2c3c329c3935/tests/nonbonded/test_nonbonded_pair_list.py#L21-L24)).

Reply:
Thanks -- that makes sense!